File: python/src/runapi/gemini_omni/resources/create_audio.py
"""Gemini Omni create-audio resource (synchronous)."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, ValidationError, RequestOptions

from ..contract_gen import CONTRACT
from ..types import CreateAudioResponse


class CreateAudio(Resource):
    """Create a reusable voice. Synchronous: ``run()`` returns the result directly."""

    ENDPOINT = "/api/v1/gemini_omni/create_audio"

    RESPONSE_CLASS = CreateAudioResponse

    MODEL = "gemini-omni-audio"

    NAME_MAX_LENGTH = 210
    VOICE_DESCRIPTION_MAX_LENGTH = 20_000
    EXAMPLE_DIALOGUE_MAX_LENGTH = 120

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a reusable voice (synchronous).

        Args:
            **params: Reusable voice parameters (audio_id, name, ...).

        Returns:
            The result.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["create-audio"], {**params, "model": self.MODEL})
        self._validate_length(params, "name", self.NAME_MAX_LENGTH)
        self._validate_length(params, "voice_description", self.VOICE_DESCRIPTION_MAX_LENGTH)
        self._validate_length(params, "example_dialogue", self.EXAMPLE_DIALOGUE_MAX_LENGTH)

    @staticmethod
    def _validate_length(params: Dict[str, Any], key: str, max_length: int) -> None:
        value = params.get(key)
        if value is None or len(str(value)) <= max_length:
            return
        raise ValidationError(f"{key} must be at most {max_length} characters")

File: ruby/lib/runapi/gemini_omni/resources/create_audio.rb
# frozen_string_literal: true

module RunApi
  module GeminiOmni
    module Resources
      # Registers a reusable voice preset from a built-in voice identity.
      # Synchronous -- only +run+ is available (no create/get polling).
      class CreateAudio
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/gemini_omni/create_audio"
        RESPONSE_CLASS = Types::CreateAudioResponse
        NAME_MAX_LENGTH = 210
        VOICE_DESCRIPTION_MAX_LENGTH = 20_000
        EXAMPLE_DIALOGUE_MAX_LENGTH = 120

        def initialize(http)
          @http = http
        end

        def run(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        private

        def validate_params!(params)
          validate_required!(params, :audio_id)
          validate_required!(params, :name)
          validate_length!(params, :name, NAME_MAX_LENGTH)
          validate_length!(params, :voice_description, VOICE_DESCRIPTION_MAX_LENGTH)
          validate_length!(params, :example_dialogue, EXAMPLE_DIALOGUE_MAX_LENGTH)
        end

        def validate_required!(params, key)
          value = param(params, key)
          return if value.is_a?(String) ? !value.empty? : !value.nil?

          raise Core::ValidationError, "#{key} is required"
        end

        def validate_length!(params, key, max_length)
          value = param(params, key)
          return if value.nil? || value.to_s.length <= max_length

          raise Core::ValidationError, "#{key} must be at most #{max_length} characters"
        end
      end
    end
  end
end
